### PR TITLE
ARROW-552: [Python] Implement getitem for DictionaryArray by returning a value from the dictionary

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

**Comment by wesm** [comment]
OS X build stalled...

File: python/pyarrow/array.pyx [hunk 1]
```diff
@@ -402,6 +402,31 @@ cdef class BinaryArray(Array):
...
 
+    cdef getitem(self, int64_t i):
+        cdef Array dictionary = self.dictionary
+        cdef int64_t index = self.indices[i].as_py()
+        return scalar.box_scalar(dictionary.type, dictionary.sp_array, index)
+
+    property dictionary:
+
+        def __get__(self):
+            cdef CDictionaryArray* darr = <CDictionaryArray*>(self.ap)
```
Comment (tebeka) [inline python/pyarrow/array.pyx]: IMO this can go in the `if` as well (same below)

Comment (xhochy) [inline python/pyarrow/array.pyx]: This is solely a cast and does will hopefully be optimised away by the compiler. I'm ok with leaving it here as Cython only allows C89 style variable definitions at the beginning of the function. Leaving it as is, is much more readable.

File: python/pyarrow/array.pxd [hunk 2]
```diff
@@ -109,7 +109,9 @@ cdef class BinaryArray(Array):
 
 
 cdef class DictionaryArray(Array):
-    pass
+    cdef:
+        object _indices, _dictionary
```
Comment (xhochy) [inline python/pyarrow/array.pxd]: Do we really need to define them? As they are Python attributes, it should be ok to omit them?

**Review by xhochy — APPROVED** [review]
+1 if we really need the `object` definitions in the `pxd`. Otherwise I would get rid of them as they are private attributes.

File: python/pyarrow/array.pyx [hunk 3]
```diff
@@ -402,6 +402,31 @@ cdef class BinaryArray(Array):
 
 cdef class DictionaryArray(Array):
 
+    cdef getitem(self, int64_t i):
+        cdef Array dictionary = self.dictionary
+        cdef int64_t index = self.indices[i].as_py()
+        return scalar.box_scalar(dictionary.type, dictionary.sp_array, index)
+
+    property dictionary:
```
Comment (tebeka) [inline python/pyarrow/array.pyx]: Why do them as property and not once in the init? It'll be less code and thread safe.

File: python/pyarrow/array.pxd (hunk 2 above)
Comment (wesm) [inline python/pyarrow/array.pxd]: I believe Cython extension types have to have all their members declared

File: python/pyarrow/array.pyx (hunk 3 above)
Comment (wesm) [inline python/pyarrow/array.pyx]: This is the @cache_readonly pattern used elsewhere. Thread safety doesn't buy us anything, and I don't see a reason to create two objects that may never be used.

File: python/pyarrow/array.pxd (hunk 2 above)
Comment (wesm) [inline python/pyarrow/array.pxd]: confirming the private members are necessary

```
E   AttributeError: 'pyarrow.array.DictionaryArray' object has no attribute '_dictionary'
```

**Comment by wesm** [comment]
rebased

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**